src/Umbraco.Tests.Integration/Umbraco.Infrastructure/Persistence/Repositories/RelationRepositoryTest.cs

```diff
@@ -250,6 +250,7 @@ public void Get_Paged_Parent_Child_Entities_With_Same_Entity_Relation()
         }
 
         [Test]
+        [UmbracoTest(Database = UmbracoTestOptions.Database.NewSchemaPerTest, Boot = true)]
```

Comment:
I can reproduce this on netcore/netcore, for which the tests pass on the CI pipeline so there might just be something fucky about this test + my pc.

Will remove the attr and see what happens.

```
System.InvalidOperationException : No data for media 1061
   at Umbraco.Web.PublishedCache.NuCache.DataSource.DatabaseDataSource.CreateMediaNodeKit(ContentSourceDto dto) in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\DataSource\DatabaseDataSource.cs:line 285
   at Umbraco.Web.PublishedCache.NuCache.DataSource.DatabaseDataSource.GetAllMediaSources(IScope scope)+MoveNext() in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\DataSource\DatabaseDataSource.cs:line 172
   at Umbraco.Web.PublishedCache.NuCache.ContentStore.SetAllFastSortedLocked(IEnumerable`1 kits, Boolean fromDb) in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\ContentStore.cs:line 702
   at Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService.LoadMediaFromDatabaseLocked(IScope scope, Boolean onStartup) in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\PublishedSnapshotService.cs:line 501
   at Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService.<LoadCachesOnStartup>b__39_3(IScope scope) in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\PublishedSnapshotService.cs:line 271
   at Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService.LockAndLoadMedia(Func`2 action) in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\PublishedSnapshotService.cs:line 477
   at Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService.LoadCachesOnStartup() in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\PublishedSnapshotService.cs:line 271
   at Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService..ctor(PublishedSnapshotServiceOptions options, IMainDom mainDom, IRuntimeState runtime, ServiceContext serviceContext, IPublishedContentTypeFactory publishedContentTypeFactory, IPublishedSnapshotAccessor publishedSnapshotAccessor, IVariationContextAccessor variationContextAccessor, IProfilingLogger profilingLogger, ILoggerFactory loggerFactory, IScopeProvider scopeProvider, IDocumentRepository documentRepository, IMediaRepository mediaRepository, IMemberRepository memberRepository, IDefaultCultureAccessor defaultCultureAccessor, IDataSource dataSource, IOptions`1 globalSettings, IEntityXmlSerializer entitySerializer, IPublishedModelFactory publishedModelFactory, UrlSegmentProviderCollection urlSegmentProviders, IHostingEnvironment hostingEnvironment, IShortStringHelper shortStringHelper, IIOHelper ioHelper, IOptions`1 config) in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\PublishedSnapshotService.cs:line 174
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Umbraco.Web.PublishedCache.NuCache.NuCacheComposer.<>c.<Compose>b__0_1(IServiceProvider factory) in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.PublishedCache.NuCache\NuCacheComposer.cs:line 32
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Umbraco.Tests.Integration.Testing.UmbracoIntegrationTest.GetRequiredService[T]() in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.Tests.Integration\Testing\UmbracoIntegrationTest.cs:line 405
   at Umbraco.Tests.Integration.Umbraco.Infrastructure.Persistence.Repositories.RelationRepositoryTest.get_RelationService() in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.Tests.Integration\Umbraco.Infrastructure\Persistence\Repositories\RelationRepositoryTest.cs:line 44
   at Umbraco.Tests.Integration.Umbraco.Infrastructure.Persistence.Repositories.RelationRepositoryTest.CreateTestDataForPagingTests(List`1& createdContent, List`1& createdMembers, List`1& createdMedia) in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.Tests.Integration\Umbraco.Infrastructure\Persistence\Repositories\RelationRepositoryTest.cs:line 321
   at Umbraco.Tests.Integration.Umbraco.Infrastructure.Persistence.Repositories.RelationRepositoryTest.Get_Paged_Child_Entities_By_Parent_Id() in C:\Users\Paul Johnson\Dev\Umbraco-CMS\src\Umbraco.Tests.Integration\Umbraco.Infrastructure\Persistence\Repositories\RelationRepositoryTest.cs:line 255
```

Reply:
I think it is a real issue, but I can look into that

Reply:
It seems to be related to the nasty fact, that `PublishedSnapshotService` are doing lots of stuff in the ctor. Maybe a solution is to divide the ctor into a real ctor and a initialize method.

Reply:
A workaround is to call  `GetRequiredService<IPublishedSnapshotService>();` as first thing in `CreateTestDataForPagingTests`. But preferable we should look into that ctor of `PublishedSnapshotService`

Reply:
>  real ctor and a initialize method.

@JimBobSquarePants will shout at you for temporal coupling :D

Reply:
I can't reproduce anymore when running full integration test suite, i can if I run that test solo.

Crazy statics somewhere?

Reply:
Can we ditch nunit, all of the test base classes and switch to xunit2 everywhere, shouldn't take long.

Reply:
It's failing because the database is not empty when `PublishedSnapshotService` is initialized. So it seems pretty consistent if I only execute the tests in `RelationRepositoryTest`

Reply:
I feel like I'm possibly going mad then, occasionally I have it pass running it solo.

Reply:
So is Boot=true not the right thing to do then, if not how come?

NuCacheComponent will resolve a IPublishedSnapshotServicebefore test runs, this is how it works at runtime, shouldn't this be what happens in the integration tests?

Reply:
@Shazwazza did you spot this bit, @bergmania has found the issue, whilst components exist Boot=true seems to be the appropriate fix

Reply:
I still think it is an issue with the ctor of `PublishedSnapshotService`. Why should the cache be build when we call the ctor? and not when the application is ready? 
I guess we could listen to `IUmbracoApplicationLifetime.ApplicationInit` and build the cache when that is called?

ctors should be cheap to call, IMO

Reply:
I guess the question there is, if "I" resolve an IPublishedSnapshotService before it is "initialized" will things break or is the cache build required for PublishedSnapshotService required for it to function.

If so we would introduce temporal coupling https://blog.ploeh.dk/2011/05/24/DesignSmellTemporalCoupling/

Can we make it so it can be resolved without risk accidental misuse.

Reply:
That makes sense.. I don't see why we should use `IPublishedSnapshotService` before the application is ready, but to be completely safe, we could introduce a `PublishedSnapshotServiceFactory`, like suggested in Seemann's article

Reply:
> I guess we could listen to IUmbracoApplicationLifetime.ApplicationInit

If we went down that route presumably with the current codebase you could also just initialize in NuCacheComponent.Initialize.

---

Is PublishedSnapshotService even meant to be under test for the RelationRepositoryTests?

Reply:
@rustybox + @bergmania I have created a new issue/discussion here https://github.com/umbraco/Umbraco-CMS/issues/9510 , I'll let the netcore team know too so we can figure out the best way forward. 👍